### PR TITLE
Add references to the grammar

### PR DESCRIPTION
**Pull request by tsandall**
The special iterator variable ("_") is not supported yet. The _ variable will
be handled by mangling the variable name while parsing the rules (which are
still to come).

### CODE REVIEW

File: opalog/parser_test.go [hunk 1]
```diff
@@ -45,6 +45,18 @@ func TestVarTerms(t *testing.T) {
...
+        reflectTerm("foo"), reflectTerm("bar"), reflectTerm("baz"),
+    })))
+    assertParseOneTerm(t, "constants 2", "foo.bar[0].baz", reflectTerm(NewRef([]*Term{
+        reflectTerm("foo"), reflectTerm("bar"), reflectTerm(0), reflectTerm("baz"),
+    })))
+    assertParseOneTerm(t, "variables", "foo.bar[0].baz[i]", reflectTerm(NewRef([]*Term{
+        reflectTerm("foo"), reflectTerm("bar"), reflectTerm(0), reflectTerm("baz"), reflectTerm(NewVar("i")),
+    })))
+}
+
```
Comment (tim-styra) [inline opalog/parser_test.go]: What about embedded refs?

foo.bar[baz[quz]]

File: opalog/ref_test.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+	assertTermNotEqual(
+		t,
+		reflectTerm(NewRef([]*Term{reflectTerm("foo"), reflectTerm("bar"), reflectTerm(NewVar("i"))})),
+		reflectTerm(NewRef([]*Term{reflectTerm("foo"), reflectTerm("baz"), reflectTerm(NewVar("i"))})),
+	)
+	assertTermNotEqual(
+		t,
+		reflectTerm(NewRef([]*Term{reflectTerm("foo"), reflectTerm("bar"), reflectTerm(NewVar("i"))})),
+		reflectTerm(NewRef([]*Term{reflectTerm("foo"), reflectTerm("bar")})),
+	)
```
Comment (tim-styra) [inline opalog/ref_test.go]: Could include a number and a boolean

File: opalog/var_test.go [hunk 3]
```diff
@@ -9,4 +9,5 @@ import "testing"
 func TestEqualVarTerms(t *testing.T) {
 	assertTermEqual(t, reflectTerm(NewVar("foo")), reflectTerm(NewVar("foo")))
 	assertTermNotEqual(t, reflectTerm(NewVar("foo")), reflectTerm(NewVar("foobar")))
+    assertTermNotEqual(t, reflectTerm(NewVar("foo")), reflectTerm(NewVar("bar")))
```
Comment (tim-styra) [inline opalog/var_test.go]: Not clear why this test is different than the one above it.

File: opalog/parser.go [hunk 4]
```diff
@@ -90,6 +90,10 @@ var g = &grammar{
 							},
```
Comment (tim-styra) [inline opalog/parser.go]: I've forgotten...why is this file checked into the repo?

File: opalog/term.go [hunk 5]
```diff
@@ -111,7 +123,11 @@ func (term1 *Term) Equal(term2 *Term) bool {
 	case VAR:
 		var1 := term1.Value.(*Var)
 		var2 := term2.Value.(*Var)
-		return var1.Name == var2.Name
+		return var1.Equal(var2)
+	case REF:
+		ref1 := term1.Value.(*Ref)
+		ref2 := term2.Value.(*Ref)
+		return ref1.Equal(ref2)
```
Comment (tim-styra) [inline opalog/term.go]: Still not sure what to think of the tradeoffs between having 1 struct for each of Var, Array, Object, etc. and having 1 struct for all Terms.  Seems that in both cases we'll want this kind of Term Equal Term function, at least for convenience.  I'd say we'll only want 1 Struct of memory overhead for each Term.  Right now it looks like we have 2: Term and Ref.  (Or does Go optimize away structs somehow?)  Maybe that's premature optimization, but it's also inevitable, so maybe it's best to just have 1 at the start.

**Comment by tsandall** [comment]
The build failure is because Travis needs to have glide/pigeon installed. I will get this fixed.

**Pull request merged**